beacon-chain/verification/initializer.go

```diff
@@ -0,0 +1,115 @@
+package verification
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain/kzg"
+	forkchoicetypes "github.com/prysmaticlabs/prysm/v4/beacon-chain/forkchoice/types"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/startup"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/interfaces"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+// Database represents the db methods that the verifiers need.
+type Database interface {
+	Block(ctx context.Context, blockRoot [32]byte) (interfaces.ReadOnlySignedBeaconBlock, error)
+}
+
+// Forkchoicer represents the forkchoice methods that the verifiers need.
+// Note that forkchoice is used here in a lock-free fashion, assuming that a version of forkchoice
+// is given that internally handles the details of locking the underlying store.
+type Forkchoicer interface {
+	FinalizedCheckpoint() *forkchoicetypes.Checkpoint
+	HasNode([32]byte) bool
+	IsCanonical(root [32]byte) bool
+	Slot([32]byte) (primitives.Slot, error)
+}
+
+// StateByRooter describes a stategen-ish type that can produce arbitrary states by their root
+type StateByRooter interface {
+	StateByRoot(ctx context.Context, blockRoot [32]byte) (state.BeaconState, error)
+}
+
+// sharedResources provides access to resources that are required by different verification types.
+// for example, sidecar verifcation and block verification share the block signature verification cache.
+type sharedResources struct {
```

Comment:
Why do all these resources need to be controlled by a single mutex ?

Reply:
Good catch on this mutex etc, it was messy. I cleaned it all up by moving the clock waiter, lock and ready boolean to `InitializerWaiter` https://github.com/prysmaticlabs/prysm/pull/13245/commits/944e6e38beceffefab4e7b759b73bd2c730a366f